src/charm.py

```diff
@@ -442,6 +446,25 @@ def _on_authorize_charm_action(self, event: ActionEvent):
             event.fail(f"Vault returned an error while authorizing the charm: {str(e)}")
             return
 
+    def _on_bootstrap_raft_action(self, event: ActionEvent):
+        """Bootstraps the raft cluster when a single node is present.
+
+        This is useful when Vault has lost quorum. The application must first
+        be reduced to a single unit.
+        """
+        if not self._api_address:
+            event.fail(message="Network bind address is not available")
+            return
+
+        try:
+            manager = RaftManager(self, self.machine, VAULT_SNAP_NAME, VAULT_STORAGE_PATH)
+            manager.bootstrap(self._node_id, self._api_address)
+        except ManagerError as e:
+            logger.error("Failed to bootstrap raft: %s", e)
+            event.fail(message=f"Failed to bootstrap raft: {e}")
+            return
+        event.set_results({"result": "Raft cluster bootstrapped successfully."})
+
```

Comment:
I'm curious what will happen if the application is not scaled to a single unit before running this. Will it fail? Will it leave the cluster in a broken state?

It would probably be a good idea to check here and fail the action if the prerequisite was not done.

What do you think?

Reply:
That is enforced by the manager itself, which will raise a `ManagerError`: https://github.com/canonical/vault-operator/blob/be4c676f89d2a5c83a7f8942c611f17af25a40f7/lib/charms/vault_k8s/v0/vault_managers.py#L1475-L1479

That being said, the `planned_units` / `leader` check was the best I could come up with and it doesn't seem quite right. I talked a bit with @saltiyazan about this too and we couldn't come up with anything better. 

If you have a better way to do it let me know.

Reply:
That is probably enough. The only other thing I can think of right now would be to check if this is unit 0 on top of that, but I do not think this is necessary.

Reply:
I don't think that would work for machine charms anyway.